### PR TITLE
Implement verification of receipt proof

### PR DESCRIPTION
**Pull request by Pasifaee**

### CODE REVIEW

File: lib/src/const.nr [hunk 1]
```diff
@@ -1,13 +1,17 @@
 /// Node type
 global BRANCH: u64 = 2; global EXTENSION: u64 = 0; global LEAF: u64 = 1;
 
-/// Constants for 32-byte long keys; TODO: Omit these when it is possible to replace constants with numeric generics.
-global KEY_LENGTH: u64 = 32; // (fixed) key length in bytes
-global NIBBLE_LENGTH: u64 = 64; // = 2*KEY_LENGTH
+/// Length of keccak256 hash.
+global HASH_LENGTH: u64 = 32;
+global HASH_NIBBLE_LENGTH: u64 = 64; // = 2 * HASH_LENGTH
```
Comment (marekkirejczyk) [inline lib/src/const.nr]: It seems we are fixing formatting here, (add missing spaces), I would separate that to another PR, make it easier to review for lib author.

File: lib/src/const.nr [hunk 2]
```diff
@@ -16,5 +20,8 @@ global MAX_STORAGE_VALUE_LENGTH: u64 = 32;
 /// Maximum number of bytes in an account state
 global MAX_ACCOUNT_STATE_LENGTH: u64 = 134;
 
+// TODO: make it generic
+global MAX_RECEIPT_RLP_LENGTH: u64 = 525; // TODO: Understand why
```
Comment (marekkirejczyk) [inline lib/src/const.nr]: I think we can skip on the second comment if we already have the first.

File: lib/src/receipt_proof.nr [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    trie_proof::TrieProof,
+    const::{MAX_TRIE_NODE_LENGTH, LEAF, HASH_LENGTH, HASH_NIBBLE_LENGTH, RLP_IDX_KEY_NIBBLE_LENGTH, MAX_RECEIPT_RLP_LENGTH},
+    utils::{memcpy, byte_value, assert_subarray},
+    node::{key_as_nibbles, verify_internal_node, verify_node_hash, resolve2}
+};
+use dep::std::hash::keccak256;
+use crate::rlp;
+
+/// Type alias for receipt proofs. Assumes value is left-padded with zeros.
+type ReceiptProof<PROOF_LEN, RECEIPT_RLP_LENGTH> = TrieProof<3, PROOF_LEN, RECEIPT_RLP_LENGTH>;
```
Comment (marekkirejczyk) [inline lib/src/receipt_proof.nr:10]: We could generalize type, as it will be the same for Receipts, Transactions, Withdraws and perhaps blobs.
```rust
type TrieProof<PROOF_LEN, RECEIPT_RLP_LENGTH> = HashedKeyTrieProof<3, PROOF_LEN, RECEIPT_RLP_LENGTH>;

type ReceiptProof<PROOF_LEN> = TrieProof<3, PROOF_LEN, RECEIPT_RLP_LENGTH>;
type TransactionProof<PROOF_LEN> = TrieProof<3, PROOF_LEN, TRANSACTION_RLP_LENGTH>;
```

Alternativelly, to avoid renaming main type in library:
```rust
type KeyValueTrieProof<PROOF_LEN, RECEIPT_RLP_LENGTH> = TrieProof<3, PROOF_LEN, RECEIPT_RLP_LENGTH>;

```

Cound be done in next PR.

**Review by marekkirejczyk — APPROVED** [review]

File: lib/src/receipt_proof.nr [hunk 4]
```diff
@@ -0,0 +1,82 @@
+use crate::{
+    trie_proof::TrieProof,
+    const::{MAX_TRIE_NODE_LENGTH, LEAF, HASH_LENGTH, HASH_NIBBLE_LENGTH, RLP_IDX_KEY_NIBBLE_LENGTH, MAX_RECEIPT_RLP_LENGTH},
+    utils::{memcpy, byte_value, assert_subarray},
+    node::{key_as_nibbles, verify_internal_node, verify_node_hash, resolve2}
+};
+use dep::std::hash::keccak256;
```
Comment (akonior) [inline lib/src/receipt_proof.nr]: Unused `keccak256` dependency.

File: lib/src/const.nr (hunk 2 above)
Comment (LogvinovLeon) [reply lib/src/const.nr]: This should be `MAX_RECEIPT_ENCODING_LENGTH` as it's RLP + tx type
